src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,678 @@
+import java.util.Objects;
+
+/**
+ * This class provides the functionality of a Key Derivation Function (KDF),
+ * which is a cryptographic algorithm for deriving additional keys from input
+ * keying material and (optionally) other data.
+ * <p>
+ * {@code KDF} objects are instantiated with the {@code getInstance} family
+ * of methods.
+ * <p>
+ * The class has two derive methods, {@code deriveKey} and {@code deriveData}.
+ * The {@code deriveKey} method accepts an algorithm {@code String} and
+ * will return a {@code SecretKey} object with the specified algorithm. The
+ * {@code deriveData} method returns a byte array of raw data.
+ * <p>
+ * API Usage Example:
+ * {@snippet lang = java:
+ *    KDF kdfHkdf = KDF.getInstance("HKDF-SHA256");
+ *
+ *    AlgorithmParameterSpec derivationSpec =
+ *             HKDFParameterSpec.ofExtract()
+ *                              .addIKM(ikm)
+ *                              .addSalt(salt).thenExpand(info, 32);
+ *
+ *    SecretKey sKey = kdfHkdf.deriveKey("AES", derivationSpec);
+ *}
+ * <br>
+ * <h2><a id="ConcurrentAccess">Concurrent Access</a></h2>
+ * Unless otherwise documented by an implementation, the methods defined in
+ * this class are not thread-safe. Multiple threads that need to access a
+ * single object concurrently should synchronize amongst themselves and
+ * provide the necessary locking. Multiple threads each manipulating separate
+ * objects need not synchronize.
+ * <br>
+ * <h2><a id="DelayedProviderSelection">Delayed Provider Selection</a></h2>
+ * If a provider is not specified when calling one of the {@code getInstance}
+ * methods, the implementation delays the selection of the provider until the
+ * {@code deriveKey} or {@code deriveData} method is called. This is called
+ * <i>delayed provider selection</i>. The primary reason this is done is to
+ * ensure that the selected provider can handle the key material that is passed
+ * to those methods - for example, the key material may reside on a
+ * hardware device that only a specific {@code KDF} provider can utilize.
+ * <p>
+ * If a provider is not specified in the {@code getInstance} method when
+ * instantiating a {@code KDF} object, the provider is selected the first time
+ * the {@code deriveKey} or {@code deriveData} method is called, and a provider
+ * is chosen that supports the parameters passed to the {@code deriveKey} or
+ * {@code deriveData} method. If the {@code getProviderName} or {@code
+ * getParameters} method is called before the {@code deriveKey} or {@code
+ * deriveData} methods, the first provider supporting the KDF algorithm and
+ * optional {@code KDFParameters} is chosen. This provider may not support
+ * the key material that is subsequently passed to the deriveKey or
+ * deriveData methods. Therefore, it is recommended not to call the {@code
+ * getProviderName} or {@code getParameters} methods until after a key
+ * derivation operation. Once a provider is selected, it cannot be changed.
```

Comment:
If I read this correctly, the first part of this paragraph is repeating the previous paragraph but with different wording, maybe the previous paragraph is left over from a previous iteration?